MaxText/torch_datasets/parquet.py

```diff
@@ -14,49 +14,113 @@
+    """Returns an iterable of data for the assigned Parquet files.
+
+    This method will be called via a `yield from` by this class's __iter__
+    function.
+    
+    Args:
+      assigned_parquet_files: The parquest files whos data should be returned as
+        an iterable.
+    
+    Returns:
+      An iterable that provides the data using a read strategy described by the
+      implementing class's name.
+    """
+    pass
 
+  def __iter__(self) -> Iterable:
+    worker_info = torch.utils.data.get_worker_info()
     if worker_info is None:
-      # Single-process data loading.
       print("Single-process data loading detected", flush=True)
-      for each_parquet_file in self.allocated_parquet_files:
-        table = pq.ParquetFile(each_parquet_file)
-        for batch in table.iter_batches(
-            batch_size=self.batch_size, columns=self.columns
-        ):
-          yield from batch.to_pylist()
+      assigned_parquet_files = self.allocated_parquet_files
     else:
-      # Multi-process data loading.
       print("Multi-process data loading detected", flush=True)
       per_worker = int(
           math.ceil(
               len(self.allocated_parquet_files) / float(worker_info.num_workers)
           )
       )
-
       worker_id = worker_info.id
       start = worker_id * per_worker
       end = min(start + per_worker, len(self.allocated_parquet_files))
+      assigned_parquet_files = self.allocated_parquet_files[start:end]
 
-      for each_parquet_file in self.allocated_parquet_files[start:end]:
-        table = pq.ParquetFile(each_parquet_file)
-        for batch in table.iter_batches(
-            batch_size=self.batch_size, columns=self.columns
+      yield from self._iter_impl(assigned_parquet_files)
+
+
+class FileParallelRandomRead(ParquetIterableDataset):
+  "File Parallel, Random Read implementation for Parquet files."""
+
+  def _iter_impl(self, assigned_parquet_files: Iterable[str]) -> Iterable:
+    """File Parallel Random Read iterator.
+
+    Some facts:
+      - Parquet metadata for all row groups is at the beginning of the file which
+        makes getting the metadata for row groups cheap.
+
+    Some implementation details:
+      - Row groups that are not a multiple of the batch size will return partial batches
+        on the last read of the row group.
+      - The reading of the row groups within a parquet file is random.
+    """
+    for single_parquest_file in assigned_parquet_files:
+      parquet_file = pq.ParquetFile(single_parquest_file, buffer_size = 0, pre_buffer = 0)
+      # Randomize the row groups
+      randomized_row_groups = self._random_iterator(range(parquet_file.num_row_groups))
+      for current_row_group in randomized_row_groups:
+        max_logging.log(f"row_group {current_row_group} has {parquet_file.metadata.row_group(current_row_group).num_rows} rows")
+        # TODO: Read batch size until all row groups have been read.
```

Comment:
Can we remove the TODO now? If this needs to be there, can we file a ticket to track it?

Reply:
Sure! Removed.